### PR TITLE
ARROW-8877: [Rust] [DataFusion] introduce CsvReadOption struct to simplify UX

### PR DESCRIPTION
**Pull request by houqp**
A new `CsvReadOptions` is introduced to capture the following CSV read configurations:

* has_header
* optional delimiter
* optional schema
* number of records to read for schema inference

See changes in `rust/datafusion/examples/csv_sql.rs` for an example on how the new interface looks like.

Initially I thought we can unify all CSV read code path using one single options struct. It turns out it's not possible. Components from low level of the stack including `arrow::csv::reader::Read`, `CsvParition`, `CsvIterator` all work under the assumption that schema has already been defined, which makes some of the fields in CsvReadOptions irrelevant.

Therefore, the newly introduced CsvReadOptions struct is only used in high level user facing APIs including: `CsvFile`, `CsvExec`, `LogicalPlanBuilder::scan_csv`, `context::register_csv`, etc.

### CODE REVIEW

**Comment by houqp** [comment]
First follow up PR for https://github.com/apache/arrow/pull/7210.

File: rust/datafusion/src/execution/physical_plan/csv.rs [hunk 1]
```diff
@@ -28,6 +28,34 @@ use arrow::csv;
...
+    /// If schema inference is run on a file with no headers, default column names
+    /// are created.
+    pub has_header: bool,
+    /// An optional column delimiter. Defaults to `b','`.
+    pub delimiter: Option<u8>,
+    /// An optional schema representing the CSV files. If None, CSV reader will try to infer it
+    /// based on data in file.
+    pub schema: Option<&'a Schema>,
+    /// number of rows to read from CSV files for schema inference if needed. Defaults to 1000.
+    pub schema_infer_read_records: Option<usize>,
```
Comment (paddyhoran) [inline rust/datafusion/src/execution/physical_plan/csv.rs:45]: Just a nit really but there are a number of defaults that are set through `default`.  But don't we need a `new` or similar to always enforce these defaults as the comments state.  For instance, you can currently create a `CsvReadOptions` where `schema_infer_read_records` is `None` and then we have to make sure we enforce the default (1000) somewhere else where it's used.

**Review by paddyhoran — APPROVED** [review]
Thanks @houqp, nice improvement.

File: rust/datafusion/src/execution/physical_plan/csv.rs (hunk 1 above)
Comment (houqp) [reply rust/datafusion/src/execution/physical_plan/csv.rs:45]: I think to enforce that, it's best to switch to builder pattern. The UX will look like this:

```rust
    ctx.register_csv(
        "aggregate_test_100",
        &format!("{}/csv/aggregate_test_100.csv", testdata),
        CsvReadOptions::new().schema_infer_read_records(50).delimiter(b'|'),
        true,
    )?;
```

What do you think?

Comment (paddyhoran) [reply rust/datafusion/src/execution/physical_plan/csv.rs:45]: Yea, I think that works.

**Review by andygrove — APPROVED** [review]
I've only skimmed through this but LGTM. Thanks @houqp

**Review by nevi-me — COMMENTED** [review]
@houqp the test failures are due to missing documentation, so LGTM subject to passing CI

**Comment by houqp** [comment]
@nevi-me sorry, the PR was in WIP state. I have now pushed all the changes including docs and removing default trait to enforce default presets.

**Comment by andygrove** [comment]
@houqp Could you push an empty commit to trigger CI. The tests are showing as skipped, probably due to WIP in title.

**Comment by houqp** [comment]
Ha, I have been wondering why those tasks are skipped, now it makes sense. @andygrove updated, all builds are passing now.

**Review by andygrove — APPROVED** [review]
This is great. Thanks for all the additional cleanup.

**Pull request closed without merging**